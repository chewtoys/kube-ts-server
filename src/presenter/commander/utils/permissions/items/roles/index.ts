import { QUERY_REGEX, UUID_REGEX } from '../../../../../../constants';
import {
  ROLES_ASSIGN_PERMISSION,
  ROLES_CREATE_ITEM,
  ROLES_DELETE_ITEM,
  ROLES_DELETE_ITEMS,
  ROLES_GET_ITEM,
  ROLES_GET_ITEMS,
  ROLES_REPLACE_ITEM,
  ROLES_REVOKE_PERMISSION,
  ROLES_UPDATE_ITEM,
} from '../../../../../../constants/permissions';

export const ROLES_URL_REGEX = '(/api/v1/roles)';

const rolesPermissions = [
  {
    method: 'GET',
    name: ROLES_GET_ITEM,
    url: `^${ROLES_URL_REGEX}/${UUID_REGEX}${QUERY_REGEX}$`,
  },
  {
    method: 'GET',
    name: ROLES_GET_ITEMS,
    url: `^${ROLES_URL_REGEX}${QUERY_REGEX}$`,
  },
  {
    method: 'POST',
    name: ROLES_CREATE_ITEM,
    url: `^${ROLES_URL_REGEX}${QUERY_REGEX}$`,
  },
  {
    method: 'POST',
    name: ROLES_ASSIGN_PERMISSION,
    url: `^${ROLES_URL_REGEX}/${UUID_REGEX}/permissions${QUERY_REGEX}$`,
  },
  {
    method: 'DELETE',
    name: ROLES_DELETE_ITEMS,
    url: `^${ROLES_URL_REGEX}${QUERY_REGEX}$`,
  },
  {
    method: 'DELETE',
    name: ROLES_DELETE_ITEM,
    url: `^${ROLES_URL_REGEX}/${UUID_REGEX}${QUERY_REGEX}$`,
  },
  {
    method: 'DELETE',
    name: ROLES_REVOKE_PERMISSION,
    url: `^${ROLES_URL_REGEX}/${UUID_REGEX}/permissions/${UUID_REGEX}${QUERY_REGEX}$`,
  },
  {
    method: 'PUT',
    name: ROLES_REPLACE_ITEM,
    url: `^${ROLES_URL_REGEX}/${UUID_REGEX}${QUERY_REGEX}$`,
  },
  {
    method: 'PATCH',
    name: ROLES_UPDATE_ITEM,
    url: `^${ROLES_URL_REGEX}/${UUID_REGEX}${QUERY_REGEX}$`,
  },
];

export default rolesPermissions;
