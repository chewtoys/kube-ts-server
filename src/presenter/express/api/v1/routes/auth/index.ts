import { Router } from 'express';
import {
  LOGIN,
  // REGISTER,
  // REMIND_PASSWORD,
  // RESET_PASSWORD,
} from '../../../../../../constants/routes';
import Config from '../../../../presenterFactory/Config';
import login from './login';
// import register from './register';
// import remindPassword from './remindPassword';
// import resetPassword from './resetPassword';

const authFactory = (config: Config): Router => {
  const router = Router();

  router.post(LOGIN, login(config));
  // router.post(REGISTER, register(config));
  // router.post(REMIND_PASSWORD, remindPassword(config));
  // router.post(RESET_PASSWORD, resetPassword(config));

  return router;
};

export default authFactory;
