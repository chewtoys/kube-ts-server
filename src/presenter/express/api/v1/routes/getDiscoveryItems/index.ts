import sendResponse from '@js-items/express/dist/utils/sendResponse';
import { toSnake } from 'convert-keys';
import { OK } from 'http-status-codes';
import Config from '../../../../presenterFactory/Config';
import catchErrors from '../../../../utils/errors/catchErrors';

const getDiscoveryItems = (config: Config) =>
  catchErrors(config, async (req, res) => {
    // TODO: provide validation for query params
    const type = req.query.type;

    let response;
    
    switch (type) {
      case 'course':
        response = await config.service.getCourseDetails({
          filter: JSON.parse(req.query.filter),
        });
        break;
      case 'homepage':
      default:
        response = await config.service.getDiscoveryItemsForHomepage({});
    }

    sendResponse({
      body: toSnake(response),
      req,
      res,
      status: OK,
    });
  });

export default getDiscoveryItems;
