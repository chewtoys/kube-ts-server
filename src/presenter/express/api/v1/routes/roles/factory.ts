import expressFactory from '@js-items/express/dist/factory';
import { ConflictingItemError } from '@js-items/foundation';
import { toCamel, toSnake } from 'convert-keys';
import _pick from 'ramda/src/pick';
import validateData from 'rulr/validateData';
import { v4 as uuid } from 'uuid';
import ConflictError from '../../../../../../utils/errors/http/ConflictError';
import hashPassword from '../../../../../../utils/helpers/auth/hashPassword';
import getVisibleUserProperties from '../../../../../../utils/helpers/model/getVisibleUserProperties';
import Config from '../../../../presenterFactory/Config';
import handleError from '../../../../utils/errors/handleError';
import rules, { schema } from '../../../../utils/schemas/users/createItem';

const usersFactory = (config: Config) =>
  expressFactory({
    beforeCreateItem: async ({ req, res }, handler) => {
      const transactionId = uuid();
      try {
        const payload: any = _pick(Object.keys(schema), req.body);
        validateData(rules)(payload);

        req.body.password = await hashPassword(req.body.password);
        req.body.createdAt = new Date();
        
        await handler({ transactionId });
      } catch (error) {
        let err = error;
        if (error instanceof ConflictingItemError) {
          err = new ConflictError(error.itemName, error.itemId);
        }

        handleError({ config, errorId: transactionId, req, res, error: err });
      }
    },
    convertDocumentIntoItem: ({ document }) => {
      // tslint:disable-next-line:no-console
      const documentCamelised: any = toCamel(document);

      return { ...documentCamelised };
    },
    convertItemIntoDocument: ({ item }) => {
      const user = getVisibleUserProperties(item);

      return toSnake(user);
    },
    defaultTransactionHandler: async ({ req, res }, handler) => {
      const transactionId = uuid();
      try {
        // tslint:disable-next-line:no-console
        req.body.password = await hashPassword(req.body.password);
        await handler({ transactionId });
      } catch (error) {
        handleError({ config, errorId: transactionId, req, res, error });
      }
    },
    service: config.service.users,
  });

export default usersFactory;
