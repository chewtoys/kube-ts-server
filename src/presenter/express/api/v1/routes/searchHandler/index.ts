import sendResponse from '@js-items/express/dist/utils/sendResponse';
import { toSnake } from 'convert-keys';
import { OK } from 'http-status-codes';
import Record from 'rulr/Record';
import validateData from 'rulr/validateData';
import { SAFE_URL_LENGTH } from '../../../../../../constants';
import String from '../../../../../../utils/validation/rules/String';
import Config from '../../../../presenterFactory/Config';
import catchErrors from '../../../../utils/errors/catchErrors';


const searchHandler = (config: Config) =>
  catchErrors(config, async (req, res) => {
    const name = config.appConfig.http.client.searchQueryParamName;

    const validationSchema = {
      [name]: String(0, SAFE_URL_LENGTH),
    };

    const rules = Record(validationSchema);
    const value = req.query[name];

    const payload: any = {[name]:value};

    validateData(rules)(payload);

    const response = await config.service.search({
      query: value
    });
       
    sendResponse({
      body: toSnake(response),
      req,
      res,
      status: OK,
    });
  });

export default searchHandler;
