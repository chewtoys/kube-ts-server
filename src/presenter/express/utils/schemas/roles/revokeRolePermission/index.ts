import Record from 'rulr/Record';
import { UUID_LENGTH } from '../../../../../../constants';
import String from '../../../../../../utils/validation/rules/String';

export const schema = {
  permission_id: String(UUID_LENGTH, UUID_LENGTH),
  role_id: String(UUID_LENGTH, UUID_LENGTH),
};

const rules = Record(schema);

export default rules;
