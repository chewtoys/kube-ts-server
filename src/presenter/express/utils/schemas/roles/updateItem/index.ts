import Record from 'rulr/Record';
import { VARCHAR_LENGTH } from '../../../../../../constants';
import Optional from '../../../../../../utils/validation/rules/Optional';
import String from '../../../../../../utils/validation/rules/String';

import baseSchema from '../base/schema';

export const beforeUpdateSchema = {
  ...baseSchema,
  name: Optional(String(0, VARCHAR_LENGTH)),
};

const rules = Record(beforeUpdateSchema);

export default rules;
