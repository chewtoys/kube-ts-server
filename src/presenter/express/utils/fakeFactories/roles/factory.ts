import { v4 as uuid } from 'uuid';
import { STUDENT } from '../../../../../constants/roles';
import Role from '../../../../../types/items/Role';
import baseFactory, { Options } from '../index';

const createRoleItemData = async () => ({
  id: uuid(),
  name: STUDENT,
});

const roleFactory = async (options: Options<Role>) => {
  const itemData = await createRoleItemData();

  return baseFactory<Role>(itemData as Partial<Role>)(options);
};

export default roleFactory;
