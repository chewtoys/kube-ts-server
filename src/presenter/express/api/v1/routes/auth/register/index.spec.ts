import dotenv from 'dotenv';
dotenv.config();
import faker from 'faker';
import { CONFLICT, CREATED } from 'http-status-codes';
import { TEXT_LENGTH } from '../../../../../../../constants';
import { STUDENT } from '../../../../../../../constants/roles';
import { API_V1, AUTH, REGISTER } from '../../../../../../../constants/routes';
import usersFactory from '../../../../../utils/fakeFactories/users/factory';
import assertOnResponseAndStatus, {
  BaseAssertionOptions,
} from '../../../../../utils/tests/assertOnResponseAndStatus';
import initTests from '../../../../../utils/tests/initTests';
import {
  TEST_DIFFERENT_VALID_PASSWORD,
  TEST_UTC_DATE,
  TEST_UUID,
  TEST_VALID_EMAIL,
  TEST_VALID_PASSWORD,
} from '../../../../../utils/tests/testData';

const emailSubject = 'subject';
const link = 'link';

jest.mock('uuid', () => ({
  v4: jest.fn(() => TEST_UUID),
}));

jest.mock('../../../../../../../utils/helpers/url/getVerifyEmailUrl', () =>
  jest.fn(() => link)
);

jest.mock('../../../../../../../utils/helpers/date/getUtcDate', () =>
  jest.fn(() => TEST_UTC_DATE)
);

jest.mock('../../../../../../../utils/helpers/auth/generateToken', () =>
  jest.fn(() => 'token')
);

const validRequiredFields = {
  email: TEST_VALID_EMAIL,
  password: TEST_VALID_PASSWORD,
  password_confirmation: TEST_VALID_PASSWORD,
};

const REGISTER_URL = `${API_V1}${AUTH}${REGISTER}`;

export const assertWithRequiredFieldAlreadyIncluded = ({
  request,
  fields,
  statusCode,
}: BaseAssertionOptions) =>
  assertOnResponseAndStatus({
    fields: {
      ...validRequiredFields,
      ...fields,
    },
    request,
    statusCode,
    url: REGISTER_URL,
  });

describe('@presenter/auth/register', () => {
  const { request, service, mailServer } = initTests({ useMailServer: true });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('fails to register user without required fields', async () => {
    await assertOnResponseAndStatus({
      request,
      url: REGISTER_URL,
    });
  });

  it('fails to register user when email is missing out of required properties', async () => {
    await assertOnResponseAndStatus({
      fields: {
        password: TEST_VALID_PASSWORD,
        password_confirmation: TEST_VALID_PASSWORD,
      },
      request,
      url: REGISTER_URL,
    });
  });

  it('fails to register user when password is missing out of required properties', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        password_confirmation: TEST_VALID_PASSWORD,
      },
      request,
      url: REGISTER_URL,
    });
  });

  it('fails to register user when password_confirmation is missing out of required properties', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        password: TEST_VALID_PASSWORD,
      },
      request,
      url: REGISTER_URL,
    });
  });

  it('fails to register user when password and password_confirmation are not matching', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        password: TEST_VALID_PASSWORD,
        password_confirmation: TEST_DIFFERENT_VALID_PASSWORD,
      },
      request,
      url: REGISTER_URL,
    });
  });

  it('fails to register user when bio exceeds number of characters', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: {
        bio: faker.random.alphaNumeric(TEXT_LENGTH + 1),
      },
      request,
    });
  });

  it('fails to register user when date_of_birth is not valid date', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: { date_of_birth: 'invalid date' },
      request,
    });
  });

  it('fails to register user when first_name is invalid', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: { first_name: 3222 },
      request,
    });
  });

  it('fails to register user when gender is invalid', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: { gender: 'non-existing-gender' },
      request,
    });
  });

  it('fails to register user when last_name is invalid', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: { last_name: 3232 },
      request,
    });
  });

  it('fails to register user when user with the email already exist', async () => {
    await usersFactory({
      overrides: {
        email: validRequiredFields.email,
      },
      service: service.users,
    });

    await assertWithRequiredFieldAlreadyIncluded({
      request,
      statusCode: CONFLICT,
    });
  });

  it('registers user returning back it visible fields and jwt token', async () => {
    jest.mock('../../../../../../../translator/factory.ts', () => () => ({
      verifyYourEmailHtml: jest.fn(() => link),
      verifyYourEmailSubject: () => emailSubject,
      verifyYourEmailText: jest.fn(() => link),
    }));

    const fields = {
      bio: 'short bio',
      date_of_birth: '1970-01-01',
      first_name: 'John',
      gender: 'male',
      last_name: 'Doe',
    };

    const { body: id } = await assertWithRequiredFieldAlreadyIncluded({
      fields,
      request,
      statusCode: CREATED,
    });

    expect(id.roles).toEqual([STUDENT]);


    const { item } = await service.users.getItem({
      id,
    });

    expect(item.verifiedAt).toBeUndefined();
    expect(item.verifyToken).toBe(TEST_UUID);

    const messages = mailServer.messages;

    expect(messages.length).toBe(1);

    const { textContent, subject, to, from, htmlContent } = messages[0];

    expect(from).toMatchSnapshot('from');
    expect(htmlContent).toMatchSnapshot('htmlContent');
    expect(textContent).toMatchSnapshot('textContent');
    expect(subject).toMatchSnapshot('subject');
    expect(to).toMatchSnapshot('to');
  });
  // tslint:disable-next-line:max-file-line-count
});
