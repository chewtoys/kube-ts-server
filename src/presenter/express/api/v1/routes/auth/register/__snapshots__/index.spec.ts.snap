// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@presenter/auth/register fails to register user when bio exceeds number of characters 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "bio",
      "message": "Expected string between 0 and 65535 characters long",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user when date_of_birth is not valid date 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "date_of_birth",
      "message": "Expected date in one of the formats: DD-MM-YYYY, YYYY-MM-DD",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user when email is missing out of required properties 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "Expected string between 0 and 255 characters long",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user when first_name is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "first_name",
      "message": "Expected string between 0 and 255 characters long",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user when gender is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "gender",
      "message": "Expected one of the following: female, male",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user when last_name is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "last_name",
      "message": "Expected string between 0 and 255 characters long",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user when password and password_confirmation are not matching 1`] = `
Object {
  "errors": Array [
    Object {
      "fields": Array [
        "password",
        "password_confirmation",
      ],
      "message": "Expected password to match password_confirmation",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user when password is missing out of required properties 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "password",
      "message": "Expected string between 0 and 255 characters long",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user when password_confirmation is missing out of required properties 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "password_confirmation",
      "message": "Expected string between 0 and 255 characters long",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user when user with the email already exist 1`] = `
Object {
  "message": "User already exists",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register fails to register user without required fields 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "Expected string between 0 and 255 characters long",
    },
    Object {
      "field": "password",
      "message": "Expected string between 0 and 255 characters long",
    },
    Object {
      "field": "password_confirmation",
      "message": "Expected string between 0 and 255 characters long",
    },
  ],
  "message": "Validation failed",
  "transactionId": "25742910-c654-4481-949f-c3f01e4e823f",
}
`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token 1`] = `
Object {
  "roles": Array [
    "student",
  ],
  "token": "token",
  "user": Object {
    "bio": "short bio",
    "created_at": "2000-04-05T23:26:42.000Z",
    "date_of_birth": "2000-04-05",
    "email": "valid.email@test.com",
    "first_name": "John",
    "gender": "male",
    "id": "25742910-c654-4481-949f-c3f01e4e823f",
    "last_name": "Doe",
  },
}
`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: from 1`] = `"noreply@gamechanger.diet"`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: htmlContent 1`] = `"<h1>Hello,</h1><p>To activate your account click <a href=\\"link\\">here</a>.</p><p>kube-ts-server</p>"`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: subject 1`] = `"[kube-ts-server] Verify your email"`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: textContent 1`] = `"Hello, to activate your account click: link. kube-ts-server"`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: to 1`] = `"valid.email@test.com"`;
